src/db/mysql/models.rs

```diff
@@ -0,0 +1,352 @@
+                    .filter(bso::collection_id.eq(&bso.collection_id))
+                    .filter(bso::id.eq(&bso.id))
+                    .set(put_bso_as_changeset(&bso))
+                    .execute(&self.conn)?;
+            } else {
+                let payload = bso.payload.as_ref().map(Deref::deref).unwrap_or_default();
+                let sortindex = bso.sortindex.unwrap_or_default();
+                let ttl = bso.ttl.map_or(DEFAULT_BSO_TTL, |ttl| ttl);
+                insert_into(bso::table)
+                    .values((
+                        bso::user_id.eq(bso.user_id as i32), // XXX:
+                        bso::collection_id.eq(&bso.collection_id),
+                        bso::id.eq(&bso.id),
+                        bso::sortindex.eq(&sortindex),
+                        bso::payload.eq(payload),
+                        bso::payload_size.eq(payload.len() as i32), // XXX:
+                        bso::modified.eq(bso.modified),
+                        bso::expiry.eq(bso.modified + ttl as i64),
+                    ))
+                    .execute(&self.conn)?;
+            }
+            self.touch_collection(bso.user_id as i32, bso.collection_id, bso.modified)?;
+            // XXX:
+            Ok(bso.modified as u64)
+        })
+    }
+
+    // XXX: limit/offset i64?
+    pub fn get_bsos_sync(
+        &self,
+        user_id: u32,
+        collection_id: i32,
+        mut ids: &[&str],
+        older: i64,
+        newer: i64,
+        sort: Sorting,
+        limit: i64,
+        offset: i64,
+    ) -> Result<results::BSOs> {
+        // XXX: ensure offset/limit/newer are valid
+
+        // XXX: should error out (400 Bad Request) when more than 100
+        // are provided (move to validation layer)
+        if ids.len() > 100 {
+            // spec says only 100 ids at a time
+            ids = &ids[0..100];
+        }
+
+        // XXX: convert to raw SQL for use by other backends
+        let mut query = bso::table
+            //.select(bso::table::all_columns())
+            .select((bso::id, bso::modified, bso::payload, bso::sortindex, bso::expiry))
+            .filter(bso::user_id.eq(user_id as i32)) // XXX:
+            .filter(bso::collection_id.eq(collection_id as i32)) // XXX:
+            .filter(bso::modified.lt(&older))
+            .filter(bso::modified.gt(&newer))
+            .filter(bso::expiry.gt(ms_since_epoch()))
+            .into_boxed();
+
+        if !ids.is_empty() {
+            query = query.filter(bso::id.eq_any(ids));
+        }
+
+        query = match sort {
+            Sorting::Index => query.order(bso::sortindex.desc()),
+            Sorting::Newest => query.order(bso::modified.desc()),
+            Sorting::Oldest => query.order(bso::modified.asc()),
+            _ => query,
+        };
+
+        // fetch an extra row to detect if there are more rows that
+        // match the query conditions
+        query = query.limit(if limit >= 0 { limit + 1 } else { limit });
```

Comment:
The docs aren't clear on what happens when `limit < 0`, does it default to "no limit"?